### PR TITLE
Update `about()` and document latest package support

### LINKED ISSUE #1201: Update `about()` and document latest package support
Issue Description
-----------------

Latest supported versions of optional dependencies are not well documented. It is unclear which frontend version(s) to install after a user installs through `pip install mitiq`, but wants to use Qiskit, Pennylane, Braket, and/or PyQuil.

Proposed Solution
-----------------

- [x] Update `about()` to include Pennylane
- [x] Create a function `latest_supported()` to parse Qiskit, Pennylane (and Pennylane-Qiskit), Braket, and PyQuil versions from `dev_requirements.txt`
- [x] Add `Latest Supported` as a column in `about()` under `Optional Dependencies`
- [x] Create a user warning when package versions are ahead of latest (from Andrea):
> e.g. something like: `WARNING: Mitiq is tested with version xxx of Qiskit. Your version is higher and may be incompatible...`
- [x] Add warning to `about()`
- [x] Add warning to `QPROGRAM` type builder
- [ ] Add `latest_supported()` with output to docs under [Supported Frontends](https://mitiq.readthedocs.io/en/stable/#supported-frontends)

Additional References
---------------------

Discussed in Mitiq meeting 04/01/22. See this [thread on discord](https://discord.com/channels/764231928676089909/958388424472166400).

### PR DESCRIPTION
**Pull request by crazy4pi314**
Fixes #1201

Description
-----------

Please explain the changes you made here.

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [ ] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [ ] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [ ] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [x] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

**Comment by crazy4pi314** [comment]
From #1201 there is still a few things left that this does not do, and I am a bit stuck on. 

> **Add warning to QPROGRAM type builder**
I am not sure the user experience we are looking for on this/where would be the right place to handle. It seems like it would be too annoying to put anytime a conversion is called, and it would already alert the user at the time of mitiq import. Maybe that is sufficient?

> **Add latest_supported() with output to docs under [Supported Frontends](https://mitiq.readthedocs.io/en/stable/#supported-frontends)**
I have exported a function for user use called `latest_supported_packages`, but code cannot be executed in the readme. If we want that, we could have a script that greps the markdown file for versions and replaces them, but that seems tedious. Anyone have other options/suggestions? I just hardcoded them for now...

File: mitiq/_about.py [hunk 1]
```diff
@@ -15,55 +15,155 @@
...
+    return LATEST_SUPPORTED_PKGS
+
+
+def check_versions() -> None:
+    """Checks that the installed versions of Mitiq's dependencies are
+    supported by the current version of Mitiq.
+    """
+
+    for name, version in INSTALLED_PKGS.items():
+        if LATEST_SUPPORTED_PKGS[name] != version:
```
Comment (andreamari) [inline mitiq/_about.py:108]: Maybe this is a too strict condition which risks to create too many warnings. 
What about if the warning is produced only if `LATEST_SUPPORTED_PKGS[name] < version` ?

Note: if we go for this weaker check, [this tool](https://stackoverflow.com/questions/11887762/how-do-i-compare-version-numbers-in-python) may be necessary to compare versions.

File: README.md [hunk 2]
```diff
@@ -87,9 +87,9 @@ We refer to any programming language you can write quantum circuits in as a *fro
...
+| [Cirq](https://quantumai.google/cirq) 0.14.0                                                                                                                                  | [Qiskit](https://qiskit.org/) 0.35.0                                                                                   | [pyQuil](https://github.com/rigetti/pyquil) 3.0.0                                                                                                       | [Braket](https://github.com/aws/amazon-braket-sdk-python) 1.17.0                                                                                                                  | [PennyLane](https://pennylane.ai/) 0.22.2                                                                                               |
```
Comment (andreamari) [inline README.md:90]: Not sure about hard-coding versions. But I am ok with it if you think it's important.

**Review by andreamari — COMMENTED** [review]
Thanks @crazy4pi314, the about() output looks is very good and clear!
The main problem is that we should access the `requirements.txt` and `dev_requirements.txt` files in a path independent way otherwise we get errors when using Mitiq from arbitrary folders.

**Comment by andreamari** [comment]
> From #1201 there is still a few things left that this does not do, and I am a bit stuck on.
> 
> > **Add warning to QPROGRAM type builder**
> > I am not sure the user experience we are looking for on this/where would be the right place to handle. It seems like it would be too annoying to put anytime a conversion is called, and it would already alert the user at the time of mitiq import. Maybe that is sufficient?

I agree the current approach is good and is sufficient.

> > **Add latest_supported() with output to docs under [Supported Frontends](https://mitiq.readthedocs.io/en/stable/#supported-frontends)**
> > I have exported a function for user use called `latest_supported_packages`, but code cannot be executed in the readme. If we want that, we could have a script that greps the markdown file for versions and replaces them, but that seems tedious. Anyone have other options/suggestions? I just hardcoded them for now...

Fine to hardcode versions or to just remove them. 
Another option is to call `latest_supported()` in a myst file of the docs that is different from the readme. But I am not sure if there is a good place for it in the current structure. Maybe somewhere in [this page](https://mitiq.readthedocs.io/en/stable/guide/frontends-backends.html)?

File: README.md (hunk 2 above)
Comment (crazy4pi314) [reply README.md:90]: It was part of the original issue suggestion, happy to pull as yeah hard-coding is a bad code smell :)

Comment (crazy4pi314) [reply README.md:90]: Rolled back

**Comment by crazy4pi314** [comment]
The mypy issues are stumping me, I think I have created a race condition with the package loading process... :(

**Comment by andreamari** [comment]
This PR got more involved than expected. I also tried to fix some aspects at some point but the situation is complicated since dev requirements are not easily accessible from `pip` versions of Mitiq (from PyPy).

Maybe we could close it in favor of a new PR which is focused on the simplified issue #1478.

**Comment by andreamari** [comment]
Based on the discussion in https://github.com/unitaryfund/mitiq/issues/1478, we can close this PR. See  https://github.com/unitaryfund/mitiq/issues/1478#issuecomment-1251454833 therein for a good summary.

**Pull request closed without merging**